### ISSUE TITLE
rect=True training may no longer produce rectangular batch tensors

### ISSUE THREAD

**Issue by youngtboy** [ultralytics/ultralytics#25008]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

It looks like `rect=True` training may no longer produce rectangular batch tensors in the current augmentation pipeline.

In older versions, such as [8.4.51](https://github.com/ultralytics/ultralytics/blob/14ea57b11969cd872f15291e5d0bdc965bdb59f7/ultralytics/data/augment.py#L2553), `v8_transforms()` created `RandomPerspective` like this:

```python
affine = RandomPerspective(
    degrees=hyp.degrees,
    translate=hyp.translate,
    scale=hyp.scale,
    shear=hyp.shear,
    perspective=hyp.perspective,
    pre_transform=None if stretch else LetterBox(new_shape=(imgsz, imgsz)),
)
```
pre_transform applies the LetterBox transformation, which uses the `rect_shape` field to resize the image to the `batch_shape` calculated in rect mode.

However, [now](https://github.com/ultralytics/ultralytics/blob/31028670f25644e79c699f5f3aeda9f2503fd1e5/ultralytics/data/augment.py#L2739), pre_transform no longer uses LetterBox. As a result, all images seem to end up being resized to `imgsz × imgsz` during training, and `rect_shape` no longer appears to be used. It seems that rect mode is effectively no longer functioning as intended.
```python
affine = RandomPerspective(
    degrees=hyp.degrees,
    translate=hyp.translate,
    scale=hyp.scale,
    shear=hyp.shear,
    perspective=hyp.perspective,
    size=(imgsz, imgsz) if not stretch else None,
)
```
Am I missing any important details? I would appreciate it if you could help clarify my confusion.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @youngtboy, thank you for your detailed report and for your interest in Ultralytics 🚀! This is an automated response to help gather the right information quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report regarding `rect=True` behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. This is especially helpful here so we can reproduce the observed batch tensor shapes, exact training command, model, dataset configuration, package version, and any relevant logs.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Can you test this?

```
pip install git+https://github.com/ultralytics/ultralytics@rect-fix
```

**Comment by youngtboy** [comment]
Thank you very much for your quick response. Your fix is to let `RandomPerspective` use `rect_shape` as its output `size`, e.g. `size=(rect_shape[1], rect_shape[0])`. This would make the final training tensor rectangular again.

However, this is not fully equivalent to the old behavior.
In the older pipeline, the flow was:

```text
load_image -> LetterBox(rect_shape) -> RandomPerspective
```
So the image was first resized+padded in `LetterBox` to the batch-specific rectangular shape, and then random affine/perspective augmentation was applied on that already letterboxed canvas.
If `RandomPerspective` only uses rect_shape as dsize, the flow becomes:
```text
load_image -> RandomPerspective(size=rect_shape)
```
This produces a rectangular output tensor, but it does not perform the same deterministic `LetterBox` resize+padding step before the affine transform. The transform center, padding behavior, and bbox/keypoint geometry can differ.

From my perspective, I tend to prefer the previous pipeline, as it better aligns with the fundamental idea behind rectangular training. However, I may have overlooked something, and I'd be grateful to hear your thoughts.

**Comment by Y-T-G** [comment]
@youngtboy It's equivalent to old approach. Even without `LetterBox`, it's resizing and padding the same way. You can compare the train_batch plots from runs folder to see the behavior.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25009](https://github.com/ultralytics/ultralytics/pull/25009), authored by @Y-T-G with contributions from @Laughing-q.

Key changes include:

- Updated `ultralytics/data/augment.py` so augmentation sizing now checks for `labels["rect_shape"]`.
- Gives rectangular batch shapes priority over the default image shape or configured augmentation size.
- Converts `rect_shape` from height-width format into width-height format for affine transform computation.
- Helps ensure `rect=True` training preserves rectangular batch tensor shapes instead of unexpectedly producing square tensors.

To test whether this resolves the issue, install the latest code from `main`:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, this fix should be included in the next official PyPI release, so waiting for the next release and upgrading with `pip install -U ultralytics` is also an option.

Please let us know whether the merged changes resolve the issue in your training setup. Thanks 🙏 for reporting this, and any further feedback is welcome if the issue persists.
